### ISSUE TITLE
[Future Enhancement] Assigning jupyter notebooks based on student credentials

### ISSUE THREAD

**Issue by balajialg** [berkeley-dsep-infra/hubs#2867]
# Summary

In a recent conversation with staff at Biology (as part of the course Bio Indicators), @ericvd-ucb and I found that they had a unique use case to which we should give more thought as we think about feature enhancement.  He suggested that they want to assign jupyter notebook with different datasets based on the students' section numbers. Currently, our short-term fix was to write a python widget that recommends the dataset file path based on the input for the section number drop-down list provided to the student.

Considering @ryanlovett's recent work on #2063 and #2856, we might have student-related details + their course affiliations through canvas authentication (Is it right? If yes, do we get section-related details too?). We could use this data to dynamically provide a user interface for instructors to assign a student's notebooks based on their canvas credentials.  Suppose we design a workflow where instructors can assign different notebooks based on various filters (sections/courses/alphabetical order of names etc.), this could potentially help instructors/course staff distribute assignments during labs in real-time. 

It requires further validation from the perspective of how many instructors are even interested in having this use case(Something, I will try to figure out at my end) and is this a problem worthy enough to invest more energy on (considering our limited bandwidth). Once we have a sufficient number of instructors who are interested, we can prioritize this request for a sprint accordingly.

# User Stories

- As an  instructor/course support staff,  I want to assign different notebooks to students based on specific filters such as section name, student name, etc.. (during labs) to meet the instructional/grading requirements

# Tasks to complete

- Evaluate the feasibility of this request from the conversation with instructors!

**Comment by felder** [comment]
Perhaps similar functionality could also be used on a per course basis for course specific libraries as mentioned in #2884

**Comment by ryanlovett** [comment]
@balajialg The Canvas API we are starting to migrate to does not have section information, only bcourse status.

https://canvas.instructure.com/doc/api/courses.html

The SIS API we are migrating from does have section information, but we are not collecting it.

**Comment by ryanlovett** [comment]
@balajialg Though there is a Sections API and we could probably collect that too.

https://canvas.instructure.com/doc/api/sections.html

**Comment by balajialg** [comment]
@ryanlovett This is fantastic! It opens up a lot of different use cases if we are able to get student section data (& data related to their group affiliation which is not currently exposed). I don't know whether you are aware of this long-pending request from D-Lab #2759. If we could fetch relevant details from canvas and assign students different notebooks/datasets + read/write drives then we can explore research lite use cases. (In the future)

**Comment by ryanlovett** [comment]
@yuvipanda Do you foresee any trouble with adding a sections scope to the CanvasOAuthenticator config? From a user perspective, I think they'd just be prompted by bcourses to accept the first time. I'm not sure if the canvas integration required that you register these scopes with the bcourses admins in some way.

**Comment by yuvipanda** [comment]
@ryanlovett Yeah, I think bcourses admin had to grant us access to any change in scopes.

**Comment by ryanlovett** [comment]
@yuvipanda Hmm, since the sections API is under /api/v1/courses/:course_id/sections/..., we might not need additional privileges. I'll play with it.

**Comment by yuvipanda** [comment]
Assuming this is to assign people different *content* (aka notebooks to open) rather than *environments*, I don't think having this in the Authenticator will be of much use, as what content is seen is mostly influenced by the nbgitpuller link being used - not something that's controlled by the authenticator at all.

**Comment by ryanlovett** [comment]
@yuvipanda Regarding the need to "assign jupyter notebook with different datasets based on the students' section numbers", the spawner can mount additional volumes to users based on course affiliation. I think this is only done when the data is very large. So if the nbgitpuller pulls down a common hw.ipynb which references large data, the spawner could be configured to mount e.g. /data/courseX/sectionY to ~/hw4data/.

If the data is not hosted on the fileserver, and/or if each section has a different .ipynb file as you say, then the gitpuller link will have to be customized.

Lastly, if the course wanted to authorize access based on section enrollment then the authenticator or spawner would need enrollment info, but I don't think that is what is being requested.

**Comment by yuvipanda** [comment]
@ryanlovett ah, true! Must students in one section be *prevented* from accessing data in another section? If that is not the case, it is easier to just mount all the data and have different notebooks read different paths. Or perhaps just this can be set as an env variable that notebook code can read to make decisions on what data to read.

**Comment by ryanlovett** [comment]
@yuvipanda Re: "Must students in one section be prevented from accessing data in another section?" I don't know what the ask is, but probably not. I agree with you about the rest.

**Comment by balajialg** [comment]
@yuvipanda @ryanlovett The problem to be solved here (based on the articulation of course support staff) is assigning different datasets based on student section numbers while minimizing the need to create many nbgitpuller links. Creating and distributing customized notebooks (as high as 20) reading different datasets using nbgitpuller link is considered a cumbersome experience for the course admin staff.

This requirement is specific to a course and the generalizability of this problem requires more user research and conversation with the teaching team.

**Comment by ryanlovett** [comment]
@balajialg Where is the data and how large is it? Is it pulled down by nbgitpuller?

A single notebook could have code that prompts the students to specify their section number:
```
# Change this to your section number.
course_section = 0
data = f'/path/to/{course_section}.csv'
```

That would be very simple and works as-is, though it requires that students know what section they're in.

One variant that @yuvipanda suggested is:
```
# Set at login. Do not change.
course_section = os.environ.get("CANVAS_SECTION_ID", 0)
data = f'/path/to/{course_section}.csv'
```
where `CANVAS_SECTION_ID` is pulled from bcourses and set by the spawner. This requires altering the CustomSpawner to get the section ID.

Both solutions require that the course staff prepares separate files named for each section. The latter requires the Canvas section which are large identifiers, e.g. 123456, 234567, 876543; as opposed to the SIS section ID, e.g. 1, 2, 3.

**Comment by balajialg** [comment]
@ryanlovett Course [GitHub page](https://github.com/ds-modules/Bio-1B) for your reference. You can check the dataset size there (small ones). They are doing what you had suggested right now which is to get the section number input from students and then recommend the relevant dataset.  Having said that, Bio 1B folks are not happy with their current workflow as they foresee student errors with inputs and are open to a future usecase where they can distribute notebooks/datasets to students in an automated way based on their section id without messing up with nbgitpuller links. 

Thank you for highlighting the second possibility. This is something we can keep in backlog as it is not time-sensitive but can revisit when a similar usecase is required by other instructors (their openness to create different files for each section)
